File: impls/py-canonicalizer/src/tacit_canon/parse.py
"""Parser for canonical text. Bytes → typed AST.

Two passes:
  1. Tokens → nested s-expression items (SList / SAtom).
  2. Items → typed AST, validating tag, arity, and child shape.

Rejections surface as ParseError (structural) or LexError (lexical).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Union

from . import ast as A
from .lex import (
    IntTok,
    LParen,
    LexError,
    RParen,
    StrTok,
    SymTok,
    Token,
    lex,
)


class ParseError(Exception):
    pass


@dataclass(frozen=True, slots=True)
class SAtom:
    tok: Token


@dataclass(frozen=True, slots=True)
class SList:
    items: tuple["SItem", ...]


SItem = Union[SAtom, SList]


def parse(data: bytes) -> A.Node:
    tokens = lex(data)
    item, pos = _read_item(tokens, 0)
    if pos != len(tokens):
        raise ParseError(f"trailing tokens after position {pos}")
    return _build(item)


def _read_item(tokens: list[Token], pos: int) -> tuple[SItem, int]:
    if pos >= len(tokens):
        raise ParseError("unexpected end of input")
    t = tokens[pos]
    if isinstance(t, LParen):
        pos += 1
        items: list[SItem] = []
        while pos < len(tokens) and not isinstance(tokens[pos], RParen):
            child, pos = _read_item(tokens, pos)
            items.append(child)
        if pos >= len(tokens):
            raise ParseError("unclosed '('")
        return SList(tuple(items)), pos + 1
    if isinstance(t, RParen):
        raise ParseError(f"unexpected ')' at token {pos}")
    return SAtom(t), pos + 1


def _build(item: SItem) -> A.Node:
    if isinstance(item, SAtom):
        tok = item.tok
        if isinstance(tok, SymTok):
            if tok.name == "pat-wild":
                return A.PatWild()
            if tok.name == "pat-var":
                return A.PatVar()
            raise ParseError(f"bare symbol {tok.name!r} is not an expression")
        raise ParseError(f"bare atom {tok!r} is not an expression")

    if not item.items:
        raise ParseError("empty list")
    head = item.items[0]
    if not (isinstance(head, SAtom) and isinstance(head.tok, SymTok)):
        raise ParseError("list head must be a tag symbol")
    tag = head.tok.name
    args = item.items[1:]
    return _build_form(tag, args)


def _build_form(tag: str, args: tuple[SItem, ...]) -> A.Node:
    match tag:
        case "lam":
            _check_arity(tag, args, 1)
            return A.Lam(_build(args[0]))
        case "app":
            _check_arity(tag, args, 2)
            return A.App(_build(args[0]), _build(args[1]))
        case "let":
            _check_arity(tag, args, 2)
            return A.Let(_build(args[0]), _build(args[1]))
        case "rec":
            if len(args) < 2:
                raise ParseError(
                    "rec requires at least 1 binding + body (ADR 0011)"
                )
            bindings = tuple(_build(a) for a in args[:-1])
            return A.Rec(bindings, _build(args[-1]))
        case "module":
            if len(args) < 1:
                raise ParseError(
                    "module requires at least 1 binding (ADR 0011)"
                )
            return A.Module(tuple(_build(a) for a in args))
        case "if":
            _check_arity(tag, args, 3)
            return A.If_(_build(args[0]), _build(args[1]), _build(args[2]))
        case "match":
            if len(args) < 2:
                raise ParseError(
                    "match requires scrutinee + at least 1 arm (ADR 0011)"
                )
            scrutinee = _build(args[0])
            arms = []
            for a in args[1:]:
                arm = _build(a)
                if not isinstance(arm, A.Arm):
                    raise ParseError("match child must be an arm")
                arms.append(arm)
            return A.Match(scrutinee, tuple(arms))
        case "arm":
            _check_arity(tag, args, 2)
            return A.Arm(_build(args[0]), _build(args[1]))
        case "record":
            if len(args) % 2 != 0:
                raise ParseError(
                    "record requires an even number of children (sym/val pairs)"
                )
            fields: list[tuple[str, A.Node]] = []
            for k_item, v_item in zip(args[0::2], args[1::2]):
                fields.append((_bare_sym(k_item, "record field"), _build(v_item)))
            return A.Record(tuple(fields))
        case "proj":
            _check_arity(tag, args, 2)
            return A.Proj(_build(args[0]), _bare_sym(args[1], "proj field"))
        case "ctor":
            if len(args) < 1:
                raise ParseError("ctor requires a name symbol")
            name = _bare_sym(args[0], "ctor name")
            sub = tuple(_build(a) for a in args[1:])
            return A.Ctor(name, sub)
        case "ann":
            _check_arity(tag, args, 2)
            return A.Ann(_build(args[0]), _build(args[1]))
        case "var":
            _check_arity(tag, args, 1)
            n_text = _bare_int(args[0], "var index")
            n = int(n_text)
            if n < 0:
                raise ParseError("var index must be >= 0")
            return A.Var(n)
        case "int":
            _check_arity(tag, args, 1)
            text = _bare_int(args[0], "int literal")
            return A.Int_(int(text))
        case "str":
            _check_arity(tag, args, 1)
            return A.Str_(_bare_str(args[0], "str literal"))
        case "sym":
            _check_arity(tag, args, 1)
            return A.Sym(_bare_sym(args[0], "sym atom"))
        case "hole":
            _check_arity(tag, args, 2)
            diag = _bare_sym(args[0], "hole diag-id")
            payload = _build(args[1])
            if not isinstance(payload, A.Str_):
                raise ParseError("hole payload must be a (str ...) node")
            return A.Hole(diag, payload)
        case "pat-ctor":
            if len(args) < 1:
                raise ParseError("pat-ctor requires a name symbol")
            name = _bare_sym(args[0], "pat-ctor name")
            sub = tuple(_build(a) for a in args[1:])
            return A.PatCtor(name, sub)
        case _:
            raise ParseError(f"unknown tag {tag!r}")


def _check_arity(tag: str, args: tuple[SItem, ...], expected: int) -> None:
    if len(args) != expected:
        raise ParseError(
            f"{tag} expects arity {expected}, got {len(args)}"
        )


def _bare_sym(item: SItem, what: str) -> str:
    if isinstance(item, SAtom) and isinstance(item.tok, SymTok):
        return item.tok.name
    raise ParseError(f"{what} must be a bare symbol")


def _bare_int(item: SItem, what: str) -> str:
    if isinstance(item, SAtom) and isinstance(item.tok, IntTok):
        return item.tok.text
    raise ParseError(f"{what} must be a decimal integer atom")


def _bare_str(item: SItem, what: str) -> str:
    if isinstance(item, SAtom) and isinstance(item.tok, StrTok):
        return item.tok.value
    raise ParseError(f"{what} must be a string atom")


__all__ = ["parse", "ParseError", "LexError"]

File: crates/tacit-canonical/src/parse.rs
//! Parser for canonical text. Bytes → typed AST.
//!
//! Two passes:
//!   1. Tokens → nested s-expression items (SList / SAtom).
//!   2. Items → typed AST, validating tag, arity, and child shape.
//!
//! Rejections surface as ParseError (structural) or LexError (lexical).

use std::fmt;

use crate::ast::Node;
use crate::lex::{lex, LexError, Token};

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ParseError {
    Structural(String),
    Lex(LexError),
}

impl fmt::Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ParseError::Structural(m) => write!(f, "{}", m),
            ParseError::Lex(e) => write!(f, "{}", e),
        }
    }
}

impl std::error::Error for ParseError {}

impl From<LexError> for ParseError {
    fn from(e: LexError) -> Self {
        ParseError::Lex(e)
    }
}

fn err<T, S: Into<String>>(msg: S) -> Result<T, ParseError> {
    Err(ParseError::Structural(msg.into()))
}

#[derive(Debug, Clone)]
enum SItem {
    Atom(Token),
    List(Vec<SItem>),
}

pub fn parse(data: &[u8]) -> Result<Node, ParseError> {
    let tokens = lex(data)?;
    let (item, pos) = read_item(&tokens, 0)?;
    if pos != tokens.len() {
        return err(format!("trailing tokens after position {}", pos));
    }
    build(&item)
}

fn read_item(tokens: &[Token], mut pos: usize) -> Result<(SItem, usize), ParseError> {
    if pos >= tokens.len() {
        return err("unexpected end of input");
    }
    match &tokens[pos] {
        Token::LParen => {
            pos += 1;
            let mut items = Vec::new();
            while pos < tokens.len() && tokens[pos] != Token::RParen {
                let (child, np) = read_item(tokens, pos)?;
                items.push(child);
                pos = np;
            }
            if pos >= tokens.len() {
                return err("unclosed '('");
            }
            Ok((SItem::List(items), pos + 1))
        }
        Token::RParen => err(format!("unexpected ')' at token {}", pos)),
        other => Ok((SItem::Atom(other.clone()), pos + 1)),
    }
}

fn build(item: &SItem) -> Result<Node, ParseError> {
    match item {
        SItem::Atom(tok) => match tok {
            Token::Sym(name) => match name.as_str() {
                "pat-wild" => Ok(Node::PatWild),
                "pat-var" => Ok(Node::PatVar),
                _ => err(format!("bare symbol {:?} is not an expression", name)),
            },
            _ => err(format!("bare atom {:?} is not an expression", tok)),
        },
        SItem::List(items) => {
            if items.is_empty() {
                return err("empty list");
            }
            let head = &items[0];
            let tag = match head {
                SItem::Atom(Token::Sym(s)) => s.as_str(),
                _ => return err("list head must be a tag symbol"),
            };
            build_form(tag, &items[1..])
        }
    }
}

fn check_arity(tag: &str, args: &[SItem], expected: usize) -> Result<(), ParseError> {
    if args.len() != expected {
        err(format!(
            "{} expects arity {}, got {}",
            tag,
            expected,
            args.len()
        ))
    } else {
        Ok(())
    }
}

fn bare_sym(item: &SItem, what: &str) -> Result<String, ParseError> {
    if let SItem::Atom(Token::Sym(s)) = item {
        Ok(s.clone())
    } else {
        err(format!("{} must be a bare symbol", what))
    }
}

fn bare_int(item: &SItem, what: &str) -> Result<String, ParseError> {
    if let SItem::Atom(Token::Int(s)) = item {
        Ok(s.clone())
    } else {
        err(format!("{} must be a decimal integer atom", what))
    }
}

fn bare_str(item: &SItem, what: &str) -> Result<String, ParseError> {
    if let SItem::Atom(Token::Str(s)) = item {
        Ok(s.clone())
    } else {
        err(format!("{} must be a string atom", what))
    }
}

fn build_form(tag: &str, args: &[SItem]) -> Result<Node, ParseError> {
    match tag {
        "lam" => {
            check_arity(tag, args, 1)?;
            Ok(Node::Lam {
                body: Box::new(build(&args[0])?),
            })
        }
        "app" => {
            check_arity(tag, args, 2)?;
            Ok(Node::App {
                fn_: Box::new(build(&args[0])?),
                arg: Box::new(build(&args[1])?),
            })
        }
        "let" => {
            check_arity(tag, args, 2)?;
            Ok(Node::Let {
                rhs: Box::new(build(&args[0])?),
                body: Box::new(build(&args[1])?),
            })
        }
        "rec" => {
            if args.len() < 2 {
                return err("rec requires at least 1 binding + body (ADR 0011)");
            }
            let mut bindings = Vec::with_capacity(args.len() - 1);
            for a in &args[..args.len() - 1] {
                bindings.push(build(a)?);
            }
            let body = build(&args[args.len() - 1])?;
            Ok(Node::Rec {
                bindings,
                body: Box::new(body),
            })
        }
        "module" => {
            if args.is_empty() {
                return err("module requires at least 1 binding (ADR 0011)");
            }
            let mut bindings = Vec::with_capacity(args.len());
            for a in args {
                bindings.push(build(a)?);
            }
            Ok(Node::Module { bindings })
        }
        "if" => {
            check_arity(tag, args, 3)?;
            Ok(Node::If {
                cond: Box::new(build(&args[0])?),
                then: Box::new(build(&args[1])?),
                else_: Box::new(build(&args[2])?),
            })
        }
        "match" => {
            if args.len() < 2 {
                return err("match requires scrutinee + at least 1 arm (ADR 0011)");
            }
            let scrutinee = build(&args[0])?;
            let mut arms = Vec::with_capacity(args.len() - 1);
            for a in &args[1..] {
                let arm = build(a)?;
                if !matches!(arm, Node::Arm { .. }) {
                    return err("match child must be an arm");
                }
                arms.push(arm);
            }
            Ok(Node::Match {
                scrutinee: Box::new(scrutinee),
                arms,
            })
        }
        "arm" => {
            check_arity(tag, args, 2)?;
            Ok(Node::Arm {
                pattern: Box::new(build(&args[0])?),
                body: Box::new(build(&args[1])?),
            })
        }
        "record" => {
            if !args.len().is_multiple_of(2) {
                return err("record requires an even number of children (sym/val pairs)");
            }
            let mut fields = Vec::with_capacity(args.len() / 2);
            let mut i = 0;
            while i < args.len() {
                let k = bare_sym(&args[i], "record field")?;
                let v = build(&args[i + 1])?;
                fields.push((k, v));
                i += 2;
            }
            Ok(Node::Record { fields })
        }
        "proj" => {
            check_arity(tag, args, 2)?;
            Ok(Node::Proj {
                record: Box::new(build(&args[0])?),
                field: bare_sym(&args[1], "proj field")?,
            })
        }
        "ctor" => {
            if args.is_empty() {
                return err("ctor requires a name symbol");
            }
            let name = bare_sym(&args[0], "ctor name")?;
            let mut sub = Vec::with_capacity(args.len() - 1);
            for a in &args[1..] {
                sub.push(build(a)?);
            }
            Ok(Node::Ctor { name, args: sub })
        }
        "ann" => {
            check_arity(tag, args, 2)?;
            Ok(Node::Ann {
                expr: Box::new(build(&args[0])?),
                type_: Box::new(build(&args[1])?),
            })
        }
        "var" => {
            check_arity(tag, args, 1)?;
            let text = bare_int(&args[0], "var index")?;
            if text.starts_with('-') {
                return err("var index must be >= 0");
            }
            let idx: u64 = text.parse().map_err(|e: std::num::ParseIntError| {
                ParseError::Structural(format!("invalid var index: {}", e))
            })?;
            Ok(Node::Var { index: idx })
        }
        "int" => {
            check_arity(tag, args, 1)?;
            let text = bare_int(&args[0], "int literal")?;
            Ok(Node::int_from_decimal(&text))
        }
        "str" => {
            check_arity(tag, args, 1)?;
            Ok(Node::Str {
                value: bare_str(&args[0], "str literal")?,
            })
        }
        "sym" => {
            check_arity(tag, args, 1)?;
            Ok(Node::Sym {
                name: bare_sym(&args[0], "sym atom")?,
            })
        }
        "hole" => {
            check_arity(tag, args, 2)?;
            let diag = bare_sym(&args[0], "hole diag-id")?;
            let payload = build(&args[1])?;
            if !matches!(payload, Node::Str { .. }) {
                return err("hole payload must be a (str ...) node");
            }
            Ok(Node::Hole {
                diag_id: diag,
                payload: Box::new(payload),
            })
        }
        "pat-ctor" => {
            if args.is_empty() {
                return err("pat-ctor requires a name symbol");
            }
            let name = bare_sym(&args[0], "pat-ctor name")?;
            let mut sub = Vec::with_capacity(args.len() - 1);
            for a in &args[1..] {
                sub.push(build(a)?);
            }
            Ok(Node::PatCtor {
                name,
                sub_patterns: sub,
            })
        }
        "pat-int" => {
            check_arity(tag, args, 1)?;
            let text = bare_int(&args[0], "pat-int literal")?;
            Ok(Node::PatInt {
                value: if text == "-0" {
                    "0".to_string()
                } else {
                    text
                },
            })
        }
        "fn-ty" => {
            check_arity(tag, args, 3)?;
            Ok(Node::FnTy {
                arg: Box::new(build(&args[0])?),
                ret: Box::new(build(&args[1])?),
                eff: Box::new(build(&args[2])?),
            })
        }
        "ty-var" => {
            check_arity(tag, args, 1)?;
            let text = bare_int(&args[0], "ty-var index")?;
            if text.starts_with('-') {
                return err("ty-var index must be >= 0");
            }
            let idx: u64 = text.parse().map_err(|e: std::num::ParseIntError| {
                ParseError::Structural(format!("invalid ty-var index: {}", e))
            })?;
            Ok(Node::TyVar { index: idx })
        }
        "forall" => {
            check_arity(tag, args, 3)?;
            let ty_text = bare_int(&args[0], "forall ty-count")?;
            let eff_text = bare_int(&args[1], "forall eff-count")?;
            let ty_count: u32 = ty_text.parse().map_err(|e: std::num::ParseIntError| {
                ParseError::Structural(format!("invalid forall ty-count: {}", e))
            })?;
            let eff_count: u32 = eff_text.parse().map_err(|e: std::num::ParseIntError| {
                ParseError::Structural(format!("invalid forall eff-count: {}", e))
            })?;
            Ok(Node::Forall {
                ty_count,
                eff_count,
                body: Box::new(build(&args[2])?),
            })
        }
        "eff-set" => {
            // Children are bare symbol atoms (effect atom names), sorted required.
            let mut atoms = Vec::with_capacity(args.len());
            for a in args {
                atoms.push(bare_sym(a, "eff-set atom")?);
            }
            Ok(Node::EffSet { atoms })
        }
        "eff-var" => {
            check_arity(tag, args, 1)?;
            let text = bare_int(&args[0], "eff-var index")?;
            if text.starts_with('-') {
                return err("eff-var index must be >= 0");
            }
            let idx: u64 = text.parse().map_err(|e: std::num::ParseIntError| {
                ParseError::Structural(format!("invalid eff-var index: {}", e))
            })?;
            Ok(Node::EffVar { index: idx })
        }
        _ => err(format!("unknown tag {:?}", tag)),
    }
}
